File: LeetCode_2023/leetcode_problems/704.二分查找.py
from typing import List


def binary_search1(nums: List[int], target: int) -> int:
    l, r = 0, len(nums) - 1

    while l <= r:
        mid = l + (r - l) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            l = mid + 1
        else:
            r = mid - 1
    return -1


sorted_list = [1, 2, 3, 4, 7, 9, 10]
print("target= 9, idx=", binary_search1(sorted_list, 9))

File: example_c++/704.二分查找.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) { // versin 1 [left, right]
        int left = 0;
        int right = nums.size() - 1;
        int mid;

        while (left <= right) {
            mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                return mid;
            }else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return -1;
    }

    int search2(vector<int>& nums, int target) { // version 2 [left, right)
        int left = 0;
        int right = nums.size() - 1;
        int mid;

        while (left < right) {
            mid = left + (right - left) / 2;
            if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        if (nums[left] == target) {
            return left;
        }

        return -1;
    }
};

int main() {
    int target = 4;
    vector<int> nums = {1, 2, 3 ,4};
    cout << "result idx: " << Solution().search(nums, target) << endl;
    cout << "result idx: " << Solution().search2(nums, target) << endl;
    return 0;
}